### ISSUE TITLE
[Inductor] torch.compile loses signbit in torch.special.log_ndtr, causing discrepancies in sign-sensitive ops (e.g., copysign, atan2)

### ISSUE THREAD

**Issue by xyt66665000** [pytorch/pytorch#187336]
### 🐛 Describe the bug
When compiling `torch.special.log_ndtr` with the `inductor` backend, the output for large positive inputs (e.g., `x = 100.0`) evaluates to `0.0`, whereas eager mode evaluates to `-0.0`. 

While `-0.0` and `0.0` evaluate to `True` under `torch.allclose`, this difference in the signbit causes discrepancies in downstream operations that are sensitive to the sign of zero (such as `torch.copysign`, `torch.atan2`, etc.). For example, using `torch.copysign` on the output yields `-1.0` in eager mode, but `1.0` in compiled mode.

### Reproduce script
```python
import torch

print("torch:", torch.__version__)
print("cuda:", torch.version.cuda)

def fn(x):
    y = torch.special.log_ndtr(x)
    return torch.copysign(torch.tensor(1.0, device=x.device, dtype=x.dtype), y)

x = torch.tensor([100.0], device="cuda", dtype=torch.float32)

y_eager = fn(x)
y_compiled = torch.compile(fn, backend="inductor", fullgraph=True)(x)
    
print("eager:", y_eager)
print("compiled:", y_compiled)
print("equal:", torch.equal(y_eager, y_compiled))
print("allclose:", torch.allclose(y_eager, y_compiled))
```

### Output logs
```text
torch: 2.14.0.dev20260612+cu130
cuda: 13.0
eager: tensor([-1.], device='cuda:0')
compiled: tensor([1.], device='cuda:0')
equal: False
allclose: False
```

### Root cause observation
The discrepancy originates from `torch.special.log_ndtr(x)` itself. If we only return `torch.special.log_ndtr(x)` for the same input, the outputs are:
* **Eager:** `tensor([-0.], device='cuda:0')`
* **Compiled:** `tensor([0.], device='cuda:0')`

### Versions

PyTorch version: 2.14.0.dev20260612+cu130
CUDA used to build PyTorch: 13.0
OS: Ubuntu 24.04.4 LTS (x86_64)
Is CUDA available: True
CUDA runtime version: 12.0.140

cc @mruberry @kshitij12345 @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

**Comment by powerofaisinstudy-debug** [comment]
Hey @xyt66665000 ,@malfet  ! I would love to work on this issue. It looks like TorchInductor's lowering/decomposition layer for `torch.special.log_ndtr` is dropping the sign bit during graph optimization, which leads to silent correctness discrepancies down the line for sign-sensitive operators.

I have my local development environment built from source and am setting up a small reproduction script with `TORCH_COMPILE_DEBUG=1` to trace the generated FX graph and code generation blocks. Could you please assign this to me?

**Comment by wasimat404** [comment]
triage  looks like a Triton codegen thing, not the decomp
poked at this a bit, sharing what i found:
doesn't repro on CPU. swapped to device="cpu" and eager + inductor both give -0.0 from log_ndtr (signbit True), copysign agrees too. so the sign loss seems CUDA/Triton-only.
the decomp is actually fine. log_ndtr lowers through the ref in torch/_refs/special/__init__.py — big x hits the log1p(-erfc(t)/2) branch. on CPU the whole chain stays correct: erfc(t) → +0.0, -(+0.0)/2 → -0.0, log1p(-0.0) → -0.0. also confirmed log1p(-0.0) keeps its signbit under torch.compile on CPU.
no intrinsic to shortcut it. common.py has cpp=calc_log_ndtr but no triton=, and libdevice doesn't expose ndtr/log_ndtr, so the Triton path has to run the decomp above.
so my guess is the -0.0 gets dropped in Triton's codegen of that branch  maybe fast-math flushing the sign of zero through the negation or log1p. don't have a CUDA box to pin the exact op tho  flagging for anyone who can check if killing sign-of-zero flushing on that branch fixes it.

**Comment by jansel** [comment]
My agent says this is fixed by triton-lang/triton#10270, which PyTorch picked up through the Triton 3.8 pin update in pytorch/pytorch#188251 (commit 530f5443d40750c4aba9c8265e48c26db57f484e). Current main preserves the signed zero in the original CUDA reproduction, so this PR is no longer needed.

**Comment by PiesAvalon** [comment]
Thanks for fixing the original `log_ndtr`/`copysign` reproducer. I rechecked
this on the current nightly. The original `copysign` case is fixed, but a
smaller in-graph `signbit` variant still returns an incorrect boolean result:

```python
import torch


def fn(x):
    return torch.signbit(torch.special.log_ndtr(x))


x = torch.tensor(
    [10.0, 20.0, 30.0, 100.0],
    device="cuda",
    dtype=torch.float32,
)

eager = fn(x)
compiled = torch.compile(
    fn, backend="inductor", fullgraph=True
)(x)

print("eager:  ", eager)
print("compiled:", compiled)

torch.testing.assert_close(compiled, eager, rtol=0, atol=0)
```

On PyTorch `2.15.0.dev20260820+cu130`
(`a69ffdbc0fa11bf7dc90221fba0abe4b106fa214`) this prints:

```text
eager:   tensor([True, True, True, True], device='cuda:0')
compiled: tensor([ True, False, False, False], device='cuda:0')
```

For the larger inputs, eager `log_ndtr` returns negative zero, so its sign bit
is `True`. Current Inductor now also preserves the negative zero when the
`log_ndtr` value itself is returned, and applying `torch.signbit` to that
returned value outside the compiled graph gives the correct result. However,
when `signbit(log_ndtr(x))` is consumed inside the same compiled graph, it
still returns `False` for those negative-zero results.

This is deterministic on CUDA float32 and does not involve a floating-point
tolerance because the observable output is boolean. It looks like the fix
resolved the original exported-value/`copysign` path but did not cover this
in-graph `signbit` path. Would it make sense to reopen this issue, or would you
prefer a separate follow-up issue for this remaining case?
